Add short-term roadmap (#211)
tigerbeetle/tigerbeetle: README.md

## TigerBeetle (under active development)

TigerBeetle is not yet production-ready. The production version of
**TigerBeetle is now under active development**. Our [DESIGN
doc](docs/DESIGN.md) provides an overview of TigerBeetle's data
structures.

> Check out short-term Roadmap below for where we're heading!

Reviewer: ```suggestion
Check out our short-term Roadmap below for where we're heading!
```

And maybe "Short-term Roadmap" could be a link.

Author: Done and done

* [demos/io_uring](./demos/io_uring), how ring buffers can eliminate kernel syscalls, reduce server hardware requirements by a factor of two, and change the way we think about event loops.
* [demos/hash_table](./demos/hash_table), how linear probing compares with cuckoo probing, and what we look for in a hash table that needs to scale to millions (and billions) of account transfers.

## Short-Term Roadmap

Done:

1. Harden the consensus layer

In progress:

2. Harden the storage layer (until this happens you'll occasionally see panics in the LSM code)

> 3. Support querying an account for its transfers (until this happens you can only query account balances and list all transfers)

Reviewer: We don't support listing all transfers. You can query accounts or transfers by id ("point queries").

Also, maybe reword "querying an account for its transfers" to "querying for transfers by their debit/credit account". "account for its transfers" sounds kind we are querying the account, and extracting its transfers.

Author: Whoops! Thanks

* [demos/io_uring](./demos/io_uring), how ring buffers can eliminate kernel syscalls, reduce server hardware requirements by a factor of two, and change the way we think about event loops.
* [demos/hash_table](./demos/hash_table), how linear probing compares with cuckoo probing, and what we look for in a hash table that needs to scale to millions (and billions) of account transfers.

## Short-Term Roadmap

Done:

> 1. Consensus layer readiness

Reviewer 2: ```suggestion
1. Consensus protocol readiness.
```

In progress:

> 2. Storage layer readiness

Reviewer 2: ```suggestion
2. Storage engine readiness.
```

>   * Until this happens you'll occasionally see panics in the LSM code.

Reviewer 2: Would be great just to explain why the panics happen. i.e. Assertion failures as we tune overtight or incorrect assertions, or as assertions detect bugs and shut TigerBeetle down safely.

We don't want people to panic when they see that TigerBeetle panics. Rather, to explain this also to build trust.

Reviewer 2: Over time, assertion failures will become more and more rare, as we increase assurance through testing.

Author: Good call, added.

>   * See [#189](https://github.com/tigerbeetledb/tigerbeetle/issues/189) for details

Reviewer 2: ```suggestion
  * See [#189](https://github.com/tigerbeetledb/tigerbeetle/issues/189) for details.
```

To use full stops consistently for all bullets.

* [demos/io_uring](./demos/io_uring), how ring buffers can eliminate kernel syscalls, reduce server hardware requirements by a factor of two, and change the way we think about event loops.
* [demos/hash_table](./demos/hash_table), how linear probing compares with cuckoo probing, and what we look for in a hash table that needs to scale to millions (and billions) of account transfers.

## Short-Term Roadmap

Done:

1. Consensus protocol readiness.

In progress:

2. Storage engine readiness.

>   * Until this happens you'll occasionally see assertion failures in

Reviewer: ```suggestion
  * Until this is implemented you'll occasionally see assertion failures in
```

    LSM code as we tune overtight or incorrect assertions, or as
    assertions detect bugs and shut TigerBeetle down safely.
  * See [#189](https://github.com/tigerbeetledb/tigerbeetle/issues/189) for details.
3. Recovery readiness.
  * See [#212](https://github.com/tigerbeetledb/tigerbeetle/issues/212) for details.

> 3. Support querying all transfers associated with an account id.

Reviewer: ```suggestion
4. Support querying all transfers associated with an account id.
```

>   * Until this happens you can only query individual accounts and transfers by id.

Reviewer: ```suggestion
  * Until this is implemented you can only query individual accounts and transfers by id.
```
